include/history_buffer.hpp

```diff
@@ -0,0 +1,223 @@
+ * It is optimised for single-threaded use and does not provide thread-safety or multi-writer/multi-reader support.
+ *
+ * Example usage:
+ * gr::history_buffer<int, 5> buffer;
+ * buffer.push_back(1);  // buffer: [1] (size: 1, capacity: 5)
+ * buffer.push_back(2);  // buffer: [1, 2] (size: 2, capacity: 5)
+ * buffer.push_back(3);  // buffer: [1, 2, 3] (size: 3, capacity: 5)
+ * buffer.push_back(4);  // buffer: [1, 2, 3, 4] (size: 4, capacity: 5)
+ * buffer.push_back(5);  // buffer: [1, 2, 3, 4, 5] (size: 5, capacity: 5)
+ * buffer.push_back(6);  // buffer: [2, 3, 4, 5, 6] (size: 5, capacity: 5)
+ *
+ * // :
+ * buffer[0];     // value: 6 - unchecked access of last/actual sample
+ * buffer[-1];    // value: 5 - unchecked access of previous sample
+ * ...
+ * buffer.at(0);  // value: 6 - checked access of last/actual sample
+ * buffer.at(-1); // value: 5 - checked access of last/actual sample
+ * ...
+ * buffer.get_span(0, 3);  // span: [4, 5, 6]
+ * buffer.get_span(1, 3);  // span: [3, 4, 5]
+ * buffer.get_span(0);     // span: [2, 3, 4, 5, 6]
+ * buffer.get_span(1);     // span: [2, 3, 4, 5]
+ */
+template<typename T, std::size_t N = std::dynamic_extent, typename Allocator = std::allocator<T>>
+class history_buffer {
+    using signed_index_type = std::make_signed_t<std::size_t>;
+    using buffer_type       = typename std::conditional_t<N == std::dynamic_extent, std::vector<T, Allocator>, std::array<T, N * 2>>;
+    using size_type   = typename std::conditional_t<N == std::dynamic_extent, std::size_t, std::size_t>;
+
+    buffer_type _buffer;
+    std::size_t _capacity = N;
+    std::size_t _write_position{ 0 };
+    std::size_t _size{ 0 };
+
+    /**
+     * @brief maps the logical index to the physical index in the buffer.
+     */
+    constexpr inline std::size_t
+    map_index(signed_index_type index) const noexcept {
+        if constexpr (N == std::dynamic_extent) {
+            return (_write_position + _capacity - 1 + index) % _capacity;
+        } else {
+            return (_write_position + N - 1 + index) % N;
+        }
+    }
+
+public:
+    constexpr explicit history_buffer() noexcept { static_assert(N != std::dynamic_extent, "need to specify capacity"); }
+
+    constexpr explicit history_buffer(size_t capacity) : _buffer(capacity * 2), _capacity(capacity) {
+        if (capacity == 0) {
+            throw std::out_of_range("capacity is zero");
+        }
+        static_assert(N == std::dynamic_extent, "incompatible fixed capacity and using capacity argument");
+    }
+
+    /**
+     * @brief Adds an element to the end expiring the oldest element beyond the buffer's capacities.
+     */
+    constexpr void
+    push_back(const T &value) noexcept {
+        _buffer[_write_position]             = value;
+        _buffer[_write_position + _capacity] = _buffer[_write_position];
+        _write_position                      = (_write_position + 1) % _capacity;
```

Comment:
`% _capacity` can be rather expensive. If there's no good solution to use a bit-mask, a branch should be cheaper. The compiler is unlikely to figure out that it can assume `_write_position < _capacity` — though you could tell it:
```c++
if (_write_position >= _capacity) std::unreachable();
```
I expect the following to be cheaper:
```c++
++_write_position;
if (_write_position == _capacity) {
  _write_position = 0;
}
```

Reply:
Before:
```
┌───────────────────────benchmark:────────────────────────┬──────┬─#N──┬─CTX-SW─┬───CPU cache misses────┬──CPU branch misses───┬─<CPU-I>─┬──min───┬──mean──┬─stddev─┬─median─┬──max───┬─total time─┬─ops/s─┐
│ circular_buffer<int>(32) - multiple producer            │ PASS │ 10  │    71  │ 1.75M / 5.00M = 35.0% │ 62.2k / 801M =  0.0% │   69.0  │ 261 ms │ 277 ms │  13 ms │ 276 ms │ 306 ms │       3  s │ 36.1M │
│ circular_buffer<int>(32) - single producer via lambda   │ PASS │ 10  │    16  │  618k / 1.45M = 42.6% │ 42.8k / 701M =  0.0% │   57.0  │ 190 ms │ 199 ms │   8 ms │ 197 ms │ 217 ms │       2  s │ 50.3M │
│ circular_buffer<int>(32) - single producer via reserve  │ PASS │ 10  │     3  │  408k / 1.14M = 35.6% │ 43.5k / 601M =  0.0% │   52.0  │ 177 ms │ 204 ms │  24 ms │ 202 ms │ 235 ms │       2  s │ 49.1M │
│ history_buffer<int>(32)                                 │ PASS │ 10  │    18  │   290k / 645k = 45.0% │ 17.5k / 200M =  0.0% │   20.0  │ 170 ms │ 175 ms │   2 ms │ 174 ms │ 180 ms │       2  s │ 57.2M │
│ history_buffer<int, 32>                                 │ PASS │ 10  │    49  │   337k / 764k = 44.1% │ 22.6k / 201M =  0.0% │   20.0  │ 181 ms │ 186 ms │   4 ms │ 184 ms │ 197 ms │       2  s │ 53.9M │
│ circular_buffer<int>(32) - no checks                    │ PASS │ 10  │     1  │  45.1k / 102k = 44.2% │ 2.75k / 300M =  0.0% │   17.0  │  23 ms │  24 ms │ 383 us │  24 ms │  24 ms │     236 ms │  423M │
│ history_buffer<int, 32>  - no checks                    │ PASS │ 10  │     3  │   109k / 407k = 26.9% │ 10.7k / 100M =  0.0% │   14.0  │ 104 ms │ 114 ms │   7 ms │ 115 ms │ 123 ms │       1  s │ 87.9M │
│ history_buffer<int>(32)  - no checks                    │ PASS │ 10  │     2  │  77.1k / 297k = 26.0% │ 9.15k / 100M =  0.0% │   14.0  │ 101 ms │ 103 ms │   1 ms │ 103 ms │ 105 ms │       1  s │ 97.2M │
└─────────────────────────────────────────────────────────┴──────┴─────┴────────┴───────────────────────┴──────────────────────┴─────────┴────────┴────────┴────────┴────────┴────────┴────────────┴───────┘
```
with the suggested changes:
```
┌───────────────────────benchmark:────────────────────────┬──────┬─#N──┬─CTX-SW─┬───CPU cache misses────┬──CPU branch misses───┬─<CPU-I>─┬──min───┬──mean──┬─stddev─┬─median─┬──max───┬─total time─┬─ops/s─┐
│ circular_buffer<int>(32) - multiple producer            │ PASS │ 10  │     2  │  564k / 2.00M = 28.3% │ 44.9k / 801M =  0.0% │   69.0  │ 225 ms │ 232 ms │   3 ms │ 234 ms │ 236 ms │       2  s │ 43.0M │
│ circular_buffer<int>(32) - single producer via lambda   │ PASS │ 10  │    28  │   332k / 979k = 33.9% │ 44.1k / 701M =  0.0% │   57.0  │ 147 ms │ 162 ms │   7 ms │ 166 ms │ 171 ms │       2  s │ 61.6M │
│ circular_buffer<int>(32) - single producer via reserve  │ PASS │ 10  │     6  │  32.2k / 230k = 14.0% │ 35.1k / 600M =  0.0% │   52.0  │ 136 ms │ 142 ms │   8 ms │ 138 ms │ 162 ms │       1  s │ 70.7M │
│ history_buffer<int>(32)                                 │ PASS │ 10  │     1  │ 11.0k / 75.6k = 14.5% │ 7.34k / 300M =  0.0% │   20.1  │  65 ms │  66 ms │ 412 us │  66 ms │  67 ms │     659 ms │  152M │
│ history_buffer<int, 32>                                 │ PASS │ 10  │     1  │  24.3k / 103k = 23.6% │ 9.61k / 300M =  0.0% │   20.1  │  66 ms │  67 ms │   2 ms │  66 ms │  71 ms │     673 ms │  149M │
│ circular_buffer<int>(32) - no checks                    │ PASS │ 10  │     1  │ 9.91k / 52.0k = 19.0% │ 2.81k / 300M =  0.0% │   17.0  │  16 ms │  18 ms │   3 ms │  17 ms │  25 ms │     184 ms │  543M │
│ history_buffer<int, 32>  - no checks                    │ PASS │ 10  │     0  │ 4.84k / 24.1k = 20.0% │ 17.7k / 200M =  0.0% │   14.0  │  11 ms │  11 ms │  56 us │  11 ms │  12 ms │     114 ms │  878M │
│ history_buffer<int>(32)  - no checks                    │ PASS │ 10  │     0  │  893  / 10.0k =  8.9% │  943  / 200M =  0.0% │   14.0  │  11 ms │  11 ms │   2 us │  11 ms │  11 ms │     114 ms │  880M │
└─────────────────────────────────────────────────────────┴──────┴─────┴────────┴───────────────────────┴──────────────────────┴─────────┴────────┴────────┴────────┴────────┴────────┴────────────┴───────┘

```
:+1: @mattkretz well spotted ... your review is worth every penny. I owe you an :ice_cream: and :coffee: !